### PR TITLE
[#2] Implementing React.createElement

### PR DESCRIPTION
**Pull request by vitkarpov**

### CODE REVIEW

**Comment by vitkarpov** [comment]
@lukebelliveau seems no interest, doesn't it? :( Anyway, I'll move on in my fork though. That's okay, never mind.

**Comment by lukebelliveau** [comment]
Hey Vitkar - to the contrary, I'm loving the work you're doing. Life have been busy for me this week, so I haven't had time to check this out and merge it in, but I'll get it done some time this weekend. I'd love to continue collaborating.

**Comment by vitkarpov** [comment]
Super cool! Now I'm sure you exist as a human being :) Sorry if it was
pushy, I didn't mean it.

Will keep in touch!

чт, 13 апр. 2017 г. в 15:40, Luke Belliveau <notifications@github.com>:

**Comment by lukebelliveau** [comment]
No worries :)!  By the way, my intention with this repo is to use for teaching/talks, and the commit history is a bit hard to follow right now, as I was figuring it out as I went.  I'd like people to be able to go through the commit log and easily see how we created this (in the same style as my [react webapp starter kit](https://github.com/lukebelliveau/react-webapp-starterkit)). I'm thinking of doing a giant rebase, or starting from scratch with more logical commits. What do you think?

**Comment by vitkarpov** [comment]
Yeah, sure! You may change the history, seems reasonable to me. Do you want to do this once at the end?

**Comment by lukebelliveau** [comment]
I already have some local changes to the history that precede your first PR, mostly squashing readme commits and editing the commit messages to provide more context. I'm hoping to force push that sometime this weekend. I may check it out on a branch first to show you before overwriting the current history.

Another related thought - in my eyes, the current history has two major stages:

(1) 831102eb866a33543c03d6d4fb624ff715d561b7, where I initially got it working
(2) 005c525644014486bfcf69b78a04ab7f95c0ac18, your refactor into DOMComponent and CompositeComponent classes, with the React injection

I think (1) is much easier for newcomers to follow, I find it relatively easy to trace the flow.  After pulling everything out into classes and injecting React into ReactDOM (2), I find it much more difficult to follow, as there are many more moving parts.  However, it is certainly closer to the architecture of the real React and ReactDOM, we can reconcile extensions of React.Component, and it sets us up to implement updates. I don't really have a solution here, just wanted to call out that observation - any thoughts?

**Comment by vitkarpov** [comment]
So, you want to make a first commit with the very straightforward approach and then make a few ones with refactoring: more modules and interactions between them, right?

**Comment by vitkarpov** [comment]
> I don't really have a solution here, just wanted to call out that observation - any thoughts?

So, keeping the history going from simple approach to the modular one is a good idea :)
